toolchain/check/check.cpp

```diff
@@ -231,14 +250,360 @@ static auto InitPackageScopeAndImports(Context& context, UnitInfo& unit_info)
+  }
+
+  // Suspend the current method scope, which is finished but still on the
+  // decl_names_stack, and push a task to leave the scope when we're parsing
+  // inline methods. Returns `true` if inline methods should be parsed
+  // immediately.
+  auto SuspendFinishedScopeAndPush(Context& context) -> bool {
+    auto method_scope_index = method_scopes.pop_back_val();
+
+    // If we've not found any inline method definitions in this scope, clean up
+    // the stack.
+    if (method_scope_index == worklist.size() - 1) {
+      context.decl_name_stack().PopScope();
+      worklist.pop_back();
+      return false;
+    }
+
+    // If we're in a nested method scope, keep track of its scope but don't
+    // parse methods now.
+    auto& enter_scope = get<EnterMethodScope>(worklist[method_scope_index]);
+    if (enter_scope.in_method_scope) {
+      // This is a nested method scope. Suspend the inner scope so we can
+      // restore it when we come to parse the method bodies.
+      enter_scope.sus_name = context.decl_name_stack().Suspend();
+
+      // Enqueue a task to leave the nested scope.
+      worklist.push_back(LeaveMethodScope{.in_method_scope = true});
+      return false;
+    }
+
+    // We're at the end of a non-nested method scope. Prepare to start parsing
+    // inline methods. Enqueue a task to leave this outer scope and end parsing
+    // inline methods.
+    worklist.push_back(LeaveMethodScope{.in_method_scope = false});
+
+    // We'll process the worklist in reverse index order, so reverse the part of
+    // it we're about to execute so we run our tasks in the order in which they
+    // were pushed.
+    std::reverse(worklist.begin() + method_scope_index, worklist.end());
+
+    // Pop the `EnterMethodScope` that's now on the end of the worklist.
+    // We stay in that scope rather than suspending then immediately
+    // resuming it.
+    CARBON_CHECK(holds_alternative<EnterMethodScope>(worklist.back()))
+        << "Unexpected task in worklist.";
+    worklist.pop_back();
+    return true;
+  }
+
+  // Pop the next task off the worklist.
+  auto Pop() -> Task { return worklist.pop_back_val(); }
+
+  // CHECK that the work list has no further work.
+  auto VerifyEmpty() {
+    CARBON_CHECK(worklist.empty() && method_scopes.empty())
+        << "Tasks left behind on worklist.";
+  }
+
+  // A worklist of parsing tasks we'll need to do later.
+  // Don't allocate any inline storage here. A Task is fairly large, so we never
+  // want this to live on the stack. Instead, we reserve space in the
+  // constructor for a fairly large number of inline method definitions.
+  llvm::SmallVector<Task, 0> worklist;
+
+  // Indexes in `worklist` of method scopes that are currently still open.
+  llvm::SmallVector<size_t> method_scopes;
+};
+}  // namespace
+
+namespace {
+// A traversal of the node IDs in the parse tree, in the order in which we need
+// to check them.
+struct NodeIdTraversal {
```

Comment:
Had you considered moving ProcessNodeIds into this too?

Reply:
We don't really have much layering here, but given the amount of state we're tracking I've been trying to split this up based on which state the different pieces need to interact with. `ProcessNodeIds` doesn't need to interact with the private members of `NodeIdTraversal`, so I've kept it separate for now. Happy to change that if you'd find that clearer or better organized.

Reply:
FWIW, I'm now wondering if the whole traversal logic should be its own .h/.cpp, but that seems out of scope for this PR. If the class were split to its own file, the function might make more sense to merge in since there's not much of an abstraction boundary.